src/deepsparse/middlewares/middleware.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Description:
+Middlwares are used to as an intermediate step for a desired function to carry out any
+ necessary logic. Ex. Logging, timing, authentication, ...
+Pipeline and Operator uses Middleware, but middlewares logic are not a core
+ functionality of Op or Pipeline. That is, Pipeline and Op can run without middlewares,
+ and their outputs should be the same using middlewares
+Lifecycle with Pipeline using Middleware:
+Pipeline -> Middleware -> Pipeline.__call__() -> Middleware
+Lifecycle with Pipeline and Ops using Middleware:
+Pipeline -> Middleware (from pipeline) -> Pipeline.__call__()
+ -> Middleware (from op) -> Op -> Middleware (from op) -> Middleware (from pipeline)
+Usage:
+Please check tests/deepsparse/v2/middleware
+"""
+
+
+import threading
+from typing import Any, Callable, Dict, Iterator, Optional, Protocol, Sequence, Type
+
+
```

Comment:
would be good in one of the class strings to provide an example of the call/send/receive flow